### ISSUE TITLE
Electron 4.0.0 uses a too recent glibc for Raspberry PI

### ISSUE THREAD

**Issue by manast** [electron/electron#16205]
* Output of `node_modules/.bin/electron --version`:
v4.0.0

* Operating System (Platform and Version):
Raspbian 2.6 (latest raspbian at the time of filling the issue).

* Output of `node_modules/.bin/electron --version` on last known working Electron version (if applicable):
V3.0.10

**Expected Behavior**
Electron 4 should work on RPI as Electron 3 does.

**Actual behavior**
Staring an electron 4 app results in this error:
```/lib/arm-linux-gnueabihf/libc.so.6: version `GLIBC_2.27' not found```
Latest Rapsbian includes GLIBC 2.23.

**To Reproduce**
Just build an electron 4.0.0 app for RPI and run it on latest raspbian.

**Comment by manast** [comment]
Ok, it seems even worse than I thought, even older versions are now delivering binaries that are not compatible with RPI. Could it be that someone made a mistake and re-built all old binaries for arm using a newer GLIBC?

**Comment by felixrieseberg** [comment]
Nobody went and rebuild all older versions 😅 When you're saying that older versions don't work either, which versions do you mean?

In addition, would you mind elaborating why updating `libc` isn't the right solution?

**Comment by manast** [comment]
@felixrieseberg yeah, I also find it strange. But going back to 3.0.10 and building with electron-builder creates a build that is dependent on glibc >= 2.27.
The reason this is bad is because raspbian only support 2.23 which makes electron incompatible with RPI...

**Comment by kwonoj** [comment]
> building with electron-builder creates a build that is dependent

What happens if running bare electron binary or packaging without electron-builder?

**Comment by izuchukwu** [comment]
I'm currently experiencing this as well. Downloading glibc 2.27 and using `LD_LIBRARY_PATH` led to the message "Segmentation fault". I'm not building using electron-builder however, and downgrading to v3.0.13 got Electron working for me.

**Comment by williamwalker88** [comment]
Can confirm this on my end as well; deploying to multiple raspberry units (3b/3b+)

```/usr/src/app/node_modules/electron/dist/electron: /lib/arm-linux-gnueabihf/libc.so.6: version `GLIBC_2.27' not found (required by /usr/src/app/node_modules/electron/dist/electron)

**Comment by willswire** [comment]
Testing with Debian Buster, Electron 4.0.0 works.  This is because Buster includes a glibc library that is compatible.

**Comment by huiyifyj** [comment]
It seems that most of Raspberry card can't run electron normally.
Hope can solve the libc dependence as soon as possible.

**Comment by bpasero** [comment]
This also seems to prevent running Electron 4.0.x on a Ubuntu 14 32bits: https://github.com/electron/electron/issues/16407

**Comment by andrei-tara** [comment]
Same issue here :(

**Comment by takashi1kun** [comment]
Same issue here

**Comment by penghzhang** [comment]
For issue [https://github.com/electron/electron/issues/16956](url), it is hard for me to upgrade my system currently, which may cuase other applications can't work well. Could you recommand a available version of electron for my system ?

**Comment by tailorvj** [comment]
Just updating here that I have run a lab to test this out and the results are:

At the moment (until Raspbian 10 is released with glibc 2.28), the only way I'm aware of to install and run a working Electron version on a Raspberry Pi running Raspbian Stretch is:

```bash
$ sudo npm install -g electron@3.0.13 --unsafe-perm=true --allow-root --arch=armv7l --platform=linux
```

In your app package.json, make sure electron 3.0.13 is required, otherwise you will get the dreaded glibc 2.27 version error. 

This means you are running a Chromium 66 engine. I hope this info helps someone. 

BTW, We are looking for testers to the upcoming FullPageOS release. It is a Rasbian distro that allows you to easily rotate web dashboards in full screen on a Raspberry Pi connect to a screen. Drop me a line on [this FullPageOS Github issue](https://github.com/guysoft/FullPageOS/issues/268) or on our [Facebook page](https://www.facebook.com/fullpagos) 

Cheers,
Tailor

**Comment by gannonprudhomme** [comment]
> Just updating here that I have run a lab to test this out and the results are:
> 
> At the moment (until Raspbian 10 is released with glibc 2.28), the only way I'm aware of to install and run a working Electron version on a Raspberry Pi running Raspbian Stretch is:
> 
> ```shell
> $ sudo npm install -g electron@3.0.13 --unsafe-perm=true --allow-root --arch=armv7l --platform=linux
> ```
> 
> In your app package.json, make sure electron 3.0.13 is required, otherwise you will get the dreaded glibc 2.27 version error.
> 
> This means you are running a Chromium 66 engine. I hope this info helps someone.
> 
> BTW, We are looking for testers to the upcoming FullPageOS release. It is a Rasbian distro that allows you to easily rotate web dashboards in full screen on a Raspberry Pi connect to a screen. Drop me a line on [this FullPageOS Github issue](https://github.com/guysoft/FullPageOS/issues/268) or on our [Facebook page](https://www.facebook.com/fullpagos)
> 
> Cheers,
> Tailor

Can confirm that this method worked for me on the Raspberry Pi 3b+ with Raspbian Lite, although I avoided the use of `sudo` due to previous errors regarding it(plus a lot of the internet saying not to).

**Comment by taoyoush** [comment]
Change the dependencies to previous Electron version, like this:
"devDependencies": {
    "electron": "^3.1.3"
Then install

**Comment by Azzaroff** [comment]
We solved it by using the 2.28 packages from Debian's buster:

https://packages.debian.org/buster/armhf/libc6/download
https://packages.debian.org/buster/armhf/libc-bin/download
https://packages.debian.org/buster/all/libc-l10n/download
https://packages.debian.org/buster/all/locales/download

Use e.g. dpkg to locally replace the raspbian packages.
We used a Raspberry Pi 2 B with Electron 4.

**Comment by tailorvj** [comment]
@Azzaroff please note, Electron 5.0.0 works fine out if the box on Raspbian. I have upgraded from 3.1.8 and npm installed without a problem. 

New security policy in 5.0.0 disables webview node integration by default though, but this can be set in BrowserWindow() settings.

Kind regards,
Tailor VJ
FullPageOS

**Comment by rdzidziguri** [comment]
IHi guys I know this is a different type of issue, but can anyone advise me on how to run the electron-packager packed application on p3 i only have x-server installed it is light version.

**Comment by evmiguel** [comment]
Hi,

We aren't using a Raspberry Pi, but we are running Electron 4.x.x on an ARM device. The current workaround we are using is:

1.  Add `deb http://http.debian.net/debian/ testing main contrib non-free` to /etc/apt/sources.list
2.  Running the following:
```
sudo apt-get update
sudo apt-get install lib6c # This will install glibc 2.28
```

Doing this, Electron version 4.x.x works on our device.

**Comment by rdzidziguri** [comment]
@evmiguel that makes sense, but would you please list the environment as well? I mean there should be X to run window in some layer, right?

- xserver-xorg
- xserver-xorg-video-fbdev
- xinit

This is all I have for now and would love to keep things as minimal as possible. Also, I would love to know if you used electron-packager to pack the app and I am not speaking about .deb, for now, i just want to know how to run result of electron-packager which i just copy pasted in my Linux device.

**Comment by evmiguel** [comment]
@rdzidziguri Sure thing. Re: the versions of packages you've listed

- xserver-xorg: 1:7.7+19
- xserver-xorg-video-fbdev: 1:0.4.4-1+b5
- xinit: 1.3.4-3+b1

Yes, we are using electron-packager to package the app.

**Comment by rdzidziguri** [comment]
@clevertonh  Thanks a lot but it is very strange actually :( looks like i have some other problem if those are all packages u have :( so no Node installed nothing else but you are still able to run it. I am glad for you guys would you please provide any run instructions? 

Here is my output of the packager and all i am doing after that is just trying to run kitchenhub in the following way.

here is my output of the packaged app
![image](https://user-images.githubusercontent.com/6897869/57953230-921f5a80-7900-11e9-8914-3e2a43b8e764.png)

and this is what i get whenever i try to run sudo xinit kitchenhub

![image](https://user-images.githubusercontent.com/6897869/57953452-25f12680-7901-11e9-92f4-b9eb5e2840ab.png)

**Comment by openbuilds-engineer** [comment]
Logged https://bugs.launchpad.net/raspbian/+bug/1831639 with Raspbian, if you all can PLEASE go add comments maybe we can get 2.27 into Rasbian (:

**Comment by plugwash** [comment]
> Logged https://bugs.launchpad.net/raspbian/+bug/1831639 with Raspbian
Closed as invalid.

Distros aren't going to change the glibc version in an existing stable release just to satisfy some third party software.

If you want to ship binaries that people can actually use then you need to decide what distro releases you want to target, check what libraries are available and set up your build environments accordingly.

**Comment by rdzidziguri** [comment]
@plugwash Sorry but do not you find this response a little bit aggressive and insulting? you are telling large community out in the world they are all idiots not being able to identify dependencies and properly using libraries.

**Comment by openbuilds-engineer** [comment]
Either way, happy to hear of alternatives that work... Buster isnt out yet
and till then I cannot deploy my Electron 4 app on Pi. So either bye bye
all my pie users? Or loose features i need to go back to 3?

**Comment by manast** [comment]
@rdzidziguri how is @plugwash aggressive? It is apps that need to adapt to existing distributions, not the other way around. I would love to ear what is the reason electron 4 depends on glibc 2.28? they have not written a singel explanation to this in this thread despite being open for months...

**Comment by groundwater** [comment]
I've locked this thread temporarily. I'm going to unlock it in 24 hours.

In short, there is an expectation that everyone correspond as in a professional work environment. Folks managing OSS are often balancing dozens of different tradeoffs. It sucks, but it's reality.

Moving forward you are absolutely welcome to discuss how a tradeoff affects your use case.

Personal attacks, aggressive language, and similar off-topic statements are not welcome. We expect you to be your best online self.

**Comment by openbuilds-engineer** [comment]
Yay, so now that the sun's out again, can I ask some pretty newb questions about this issue:

1.  Why when I compile my app on the Pi, and then try running it, does it ask for glibc 2.27. 
2.  Is there a switch I can use during the electron-build to have it rebuild whatever was prebuilt using glibc 2.27 using glibc2.24 that is on the Pi?
3.  What are my other alternatives: I cannot downgrade Electron to 3.x (that works on the Pi)

Debian Buster will have 2.27 but its still a couple months away from release

Open to any options, just got to get my userbase happy again, so ANY plans to try will be worth it. 

Honestly I don't understand the root of the issue, where the "need glibc 2.27" error even comes from (:

**Comment by plugwash** [comment]
Note: I am the maintainer of Raspbian, I did not know "electron" existed until someone came over and filed a bug in our bugtracker. Sorry if I came across as over-agressive.

> Honestly I don't understand the root of the issue, where the "need glibc 2.27" error even comes from 

Because of the way glibc does symbol versioning and macro/inline implementations, building in an environment with a newer glibc will often (though not always) result in binaries that depend on symbols or symbol versions that are only in the newer glibc.

Most likely some piece of build infrastructure was upgraded to a newer glibc and is now cranking out binaries that won't work on systems with older glibc. The soloution would presumablly be to build electron yourself in the environment you plan to run it in, but I have no idea how involved that process is.

I just had a quick poke through the repository and it looks like the instructions for building electron yourself are at https://github.com/electron/electron/blob/master/docs/development/build-instructions-linux.md

**Comment by rdzidziguri** [comment]
@plugwash 

> The solution would presumably be to build electron yourself in the environment you plan to run it in, but I have no idea how involved that process is.

That's a reasonable solution, for now, the only issue people might face with is CI/CD pipelines (I am not mentioning slow build time on the PI itself if you mean that. but you are absolutely right. It is just a matter of time We all understand it will be fixed eventually but because PI is evolving and Electron is a really perfect solution for some advanced commercial Kiosk or Automation projects. some of those projects were pushed back because of this. There is an even newer version of Electron which is full major version upgrade and I think someone should try to upgrade to this version and see if the same issue observed.

**Comment by develar** [comment]
If you use electron-builder — since 21.0.6 any electron version from 2 is supported and you can build snap format for Raspberry PI on any OS. More info in [release notes](https://github.com/electron-userland/electron-builder/releases/tag/v21.0.6). Snap produced by electron-builder doesn't depend on system glibc version.

**Comment by tarruda** [comment]
This issue also puzzled me for the past few months, though for me it happened when I try to build electron locally on desktop.  I believe I've finally found the root cause.

TLDR

This is caused by building linux version of electron with broken debian sysroot image. Uploading a fixed sysroot image to s3 should fix it. Just run the sysroot-creator.sh (the electron version) again with the PATH env var set in a way it will use an updated version of the `nm` utility. A chromium/electron checkout should have a copy of the utility, so something like this should work:
```
export PATH=CHROMIUM_SRC_DIR/third_party/binutils/Linux_x64/Release/bin:${PATH}"
sysroot-creator-sid.sh UpdatePackageListsAmd64
sysroot-creator-sid.sh BuildSysrootAmd64
```

(the above commands need to be repeated for every supported linux architecture.

Here's the full story:

Last year chromium started using debian sid as the sysroot for building on linux, and this sysroot has a too recent glibc. In order to not lose compatibility with older distributions, the sysroot-creator.sh script applies some hacks to the sysroot to artificially reduce the required glibc version.

The function which does this is `HacksAndPatchesCommon` and it modifies some system headers (math.h and glob.h) through another script, but it needs to extract the symbols versions from libm and libc. The extraction of the symbols is done through the binutils utility `nm`, with `nm -D --defined-only --with-symbol-version` command. The problem is that the `--with-symbol-versions` is a relatively new command-line switch (it is not available in the binutils version shipped with ubuntu 16.04, for example). 

I think this is an approximation of happened to cause the broken sysroot:

- Electron CI was using ubuntu 16.04;
- When it upgraded to the chromium version used in electron 4, ubuntu 16.04 was used to create a broken debian_sysroot since the `nm` command failed(the image creation continues normally though).
- When first building electron 4 on the CI with the broken sysroot, the probably were failures due to the broken glibc compatibility, so the CI was upgraded to Ubuntu 18.04.

**Comment by MarshallOfSound** [comment]
@tarruda this is where we generate our sysroot images, what change do you think needs to be done there?

**Comment by tarruda** [comment]
@MarshallOfSound TBH I didn't use https://github.com/electron/debian-sysroot-image-creator to create my sysroot, instead I patched sysroot-creator.sh/sysroot-creator-sid.sh on chromium src checked out by electron and used that to build my sysroot image.

I did that last year when I first discussed the problem with @nornagon, but now that I found the cause I believe it was unnecessary. I think the only thing that needs to be done is ensure the sysroot script is executed in a recent enough distribution such as debian 9 or ubuntu 18.04.

I will try building a sysroot image using https://github.com/electron/debian-sysroot-image-creator and will post back the results here.

**Comment by nornagon** [comment]
d-s-i-c builds on stretch: https://github.com/electron/debian-sysroot-image-creator/blob/sid/.circleci/config.yml#L6

**Comment by electron-triage** [comment]
The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

If this is still reproducible on a supported version, please open a new issue with any other new information that a maintainer should know.

Thank you for taking the time to report this issue and helping to make Electron better! Your help is appreciated.

**Comment by gaoxiong523** [comment]
> This issue also puzzled me for the past few months, though for me it happened when I try to build electron locally on desktop. I believe I've finally found the root cause.
> 
> TLDR
> 
> This is caused by building linux version of electron with broken debian sysroot image. Uploading a fixed sysroot image to s3 should fix it. Just run the sysroot-creator.sh (the electron version) again with the PATH env var set in a way it will use an updated version of the `nm` utility. A chromium/electron checkout should have a copy of the utility, so something like this should work:
> 
> ```
> export PATH=CHROMIUM_SRC_DIR/third_party/binutils/Linux_x64/Release/bin:${PATH}"
> sysroot-creator-sid.sh UpdatePackageListsAmd64
> sysroot-creator-sid.sh BuildSysrootAmd64
> ```
> 
> (the above commands need to be repeated for every supported linux architecture.
> 
> Here's the full story:
> 
> Last year chromium started using debian sid as the sysroot for building on linux, and this sysroot has a too recent glibc. In order to not lose compatibility with older distributions, the sysroot-creator.sh script applies some hacks to the sysroot to artificially reduce the required glibc version.
> 
> The function which does this is `HacksAndPatchesCommon` and it modifies some system headers (math.h and glob.h) through another script, but it needs to extract the symbols versions from libm and libc. The extraction of the symbols is done through the binutils utility `nm`, with `nm -D --defined-only --with-symbol-version` command. The problem is that the `--with-symbol-versions` is a relatively new command-line switch (it is not available in the binutils version shipped with ubuntu 16.04, for example).
> 
> I think this is an approximation of happened to cause the broken sysroot:
> 
> * Electron CI was using ubuntu 16.04;
> * When it upgraded to the chromium version used in electron 4, ubuntu 16.04 was used to create a broken debian_sysroot since the `nm` command failed(the image creation continues normally though).
> * When first building electron 4 on the CI with the broken sysroot, the probably were failures due to the broken glibc compatibility, so the CI was upgraded to Ubuntu 18.04.

Hello, I tried to run redhat6.9 32-bit system, but I had the same problem. Could you tell me how to solve this problem?I can't upgrade my system.thanks
